core/ledger/ledgermgmt/ledger_mgmt.go

```diff
@@ -31,10 +31,12 @@ var ErrLedgerMgmtNotInitialized = errors.New("ledger mgmt should be initialized
 
 // LedgerMgr manages ledgers for all channels
 type LedgerMgr struct {
-	lock               sync.Mutex
-	openedLedgers      map[string]ledger.PeerLedger
-	ledgerProvider     ledger.PeerLedgerProvider
-	ebMetadataProvider MetadataProvider
+	lock                sync.Mutex
+	underCreationLock   sync.Mutex
+	openedLedgers       map[string]ledger.PeerLedger
+	underCreationLedger string
```

Comment:
Is it possible to use an atomic flag and `CompareAndSwap()` from [atomic pkg](https://golang.org/pkg/sync/atomic/) rather than the `underCreationLock` and `underCreationLedger`?

Reply:
@cendhu The `underCreationLedger` is not simple compare and swap. For CreateLedger (by genesisblock), it does not swap and also holds the `underCreationLock` longer so that another CreateLedgerBySnapshot cannot acquire `lock` before `CreateLedger` is done.

Reply:
The usage of `underCreationLedger` is not a simple compare and swap case. As Manish pointed out, we don't need 2 locks. I am going to remove `underCreationLock`.

Reply:
1. The `lock` mutex is used to synchronize between `Create()`, `Open()`, and `Close()`. The current semantics is that only one operation is allowed at a time. If we are planning to remove `underCreationLock`, instead use `lock` to set the status flag and release the `lock` immediately while the ledger opening process is still running in the background, we would be diverging from the current semantics. I am not saying that we should stick to the existing semantics but it is necessary to analyze the impact. 
2. Given that the `underCreationLock` is mainly used to set a flag (whether a channel creation is under-progress or not), I doubt why an atomic variable is not possible here while still retaining the existing semantics. With this, all exported methods on the ledgerMgr would check the status of the flag and inform the user about the long wait time (i.e.,  the time taken to bootstrap from the snapshot). For both checking the status and setting the status, I think, we can use atomic variable if we want to retain the existing semantics. Correct me if I have misunderstood anything.

Reply:
Discussed with @manish-sethi and @cendhu and agreed to keep the semantic and use 2 locks. Using atomic var may miss a corner case where CreateLedger and CreateLedgerSnapshot calls both pass the atomic var checking and have to compete LedgerMgr.lock.